### PR TITLE
DevKit updates for 4.x branch

### PR DESCRIPTION
**Pull request by SonataCI**
## Changelog

```markdown
### Added
- Generics to ProxyQuery and ProxyQueryInterface
```

### CODE REVIEW

File: src/Datagrid/ProxyQuery.php [hunk 1]
```diff
@@ -57,9 +60,6 @@ public function __clone()
         $this->queryBuilder = clone $this->queryBuilder;
     }
 
-    /**
-     * @return Iterator<object>
```
Comment (jordisala1991) [inline src/Datagrid/ProxyQuery.php:61]: wdyt about this alternative @VincentLanglet ? the typehint for execute is already on the parent and it is already iterable<T> so I don't think we need to add this.

Comment (VincentLanglet) [reply src/Datagrid/ProxyQuery.php:61]: That's a solution until https://github.com/doctrine/mongodb-odm/pull/2457 indeed.

File: tests/Datagrid/ProxyQueryTest.php [hunk 2]
```diff
@@ -119,8 +119,8 @@ public function testExecuteAllowsSorting(): void
         $proxyQuery->setSortBy([], ['fieldName' => 'name']);
         $proxyQuery->setSortOrder('DESC');
 
-        /** @var array{array{name: string}} $result */
-        $result = $proxyQuery->execute()->toArray();
+        /** @var iterable<array{name: string}> $result */
```
Comment (VincentLanglet) [inline tests/Datagrid/ProxyQueryTest.php:122]: If you typehint the ProxyQuery as `ProxyQuery<array{name: string}>` you won't have to typehint the result.

File: tests/Datagrid/ProxyQueryTest.php [hunk 3]
```diff
@@ -143,23 +143,26 @@ public function testExecuteAllowsSortingWithEmbedded(): void
         $proxyQuery->setSortBy([['fieldName' => 'embeddedDocument']], ['fieldName' => 'position']);
         $proxyQuery->setSortOrder('DESC');
 
-        /** @var array{array{name: string}} $result */
-        $result = $proxyQuery->execute()->toArray();
+        /** @var iterable<array{name: string}> $result */
```
Comment (VincentLanglet) [inline tests/Datagrid/ProxyQueryTest.php:146]: same

**Comment by VincentLanglet** [comment]
The changelog from https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/pull/778 should be added to this PR

**Comment by jordisala1991** [comment]
Done

**Review by VincentLanglet — APPROVED** [review]

**Pull request merged**